src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -81,7 +85,7 @@ internal virtual bool OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerO
         // Provide a default implementation for value converters.
         internal virtual bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, [MaybeNullWhen(false)] out T value)
         {
-            value = Read(ref reader, typeToConvert, options);
+            value = Read(ref reader, typeToConvert, options)!;
```

Comment:
Is the `[MaybeNullWhen(false)]` on `out T value` still correct?  I'm asking because we're returning true and suppressing the nullable warning here... are there any cases where Read could return null for a non-nullable T?  If yes, that attribute should be changed to be `[MaybeNull]` rather than `[MaybeNullWhen(false)]`

Reply:
Read can't return `null` for non-nullable types. However, the user converter could return `null` arbitrarily for reference & nullable value types. I believe `MaybeNull` should be used.

Reply:
This `virtual` implementation is actually never called, as currently only converters for ClassType.Value types (user converters & simple types, where we have all the JSON before we call the converter) would not override this. Same goes for `OnTryWrite`. @steveharter, should these throw `NotImplementedException` for now while the converter model for objects and collections is still internal?

Reply:
> should these throw NotImplementedException for now while the converter model for objects and collections is still internal?

I prefer keeping the implementation as-is for now, but could go either way short-term.

The reason it is not called is because there is an explicit optimization in `JsonConverter<T>.TryRead` that avoids the extra indirection through this method for value converters. Otherwise the method implementation is valid -- if we remove that optimization or make OnTryRead protected or public to enable a single method that can call any converter then this will be called.

Reply:
Changing the annotation to `[MaybeNull]` is valid -- this method will always return true (since value converters don't support continuation) so the original `MaybeNullWhen(false)` was not quite correct (but did no harm).

Reply:
(Yup, I realize that we don't call it to avoid the extra indirection for those cases) Makes sense to leave as-is. I brought it up because I don't think we'll be doing this in the short-term:

> if we remove that optimization or make OnTryRead protected or public to enable a single method that can call any converter then this will be called.